### ISSUE TITLE
NULL error on policy violations

### ISSUE THREAD

**Issue by BlythMeister** [DependencyTrack/dependency-track#2184]
### Current Behavior

```
2022-11-23 16:25:17,209 ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread
javax.jdo.JDOUserException: One or more instances could not be deleted
	at org.datanucleus.api.jdo.JDOPersistenceManager.deletePersistentAll(JDOPersistenceManager.java:819)
	at alpine.persistence.AbstractAlpineQueryManager.delete(AbstractAlpineQueryManager.java:484)
	at org.dependencytrack.persistence.PolicyQueryManager.reconcilePolicyViolations(PolicyQueryManager.java:168)
	at org.dependencytrack.persistence.QueryManager.reconcilePolicyViolations(QueryManager.java:533)
	at org.dependencytrack.policy.PolicyEngine.evaluate(PolicyEngine.java:98)
	at org.dependencytrack.policy.PolicyEngine.evaluate(PolicyEngine.java:69)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.performPolicyEvaluation(VulnerabilityAnalysisTask.java:130)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.inform(VulnerabilityAnalysisTask.java:76)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:101)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.datanucleus.exceptions.NucleusDataStoreException: Update of object with id "org.dependencytrack.model.ViolationAnalysis:1955" using statement "UPDATE VIOLATIONANALYSIS SET POLICYVIOLATION_ID=? WHERE ID=?" failed : com.microsoft.sqlserver.jdbc.SQLServerException: Cannot insert the value NULL into column 'POLICYVIOLATION_ID', table '15B-PRD-DependencyTrack.dbo.VIOLATIONANALYSIS'; column does not allow nulls. UPDATE fails.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:265)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1673)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:620)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:540)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7627)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3912)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:268)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:242)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeUpdate(SQLServerPreparedStatement.java:486)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at org.datanucleus.store.rdbms.SQLController.executeStatementUpdate(SQLController.java:430)
	at org.datanucleus.store.rdbms.request.UpdateRequest.execute(UpdateRequest.java:467)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.updateObjectInTable(RDBMSPersistenceHandler.java:447)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.updateObject(RDBMSPersistenceHandler.java:421)
	at org.datanucleus.store.rdbms.mapping.java.PersistableMapping.preDelete(PersistableMapping.java:1104)
	at org.datanucleus.store.rdbms.request.DeleteRequest.execute(DeleteRequest.java:213)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObjectFromTable(RDBMSPersistenceHandler.java:534)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObject(RDBMSPersistenceHandler.java:506)
	at org.datanucleus.state.StateManagerImpl.internalDeletePersistent(StateManagerImpl.java:1260)
	at org.datanucleus.state.StateManagerImpl.deletePersistent(StateManagerImpl.java:5513)
	at org.datanucleus.ExecutionContextImpl.deleteObjectInternal(ExecutionContextImpl.java:2294)
	at org.datanucleus.ExecutionContextImpl.deleteObjectWork(ExecutionContextImpl.java:2220)
	at org.datanucleus.ExecutionContextImpl.deleteObjects(ExecutionContextImpl.java:2116)
	at org.datanucleus.api.jdo.JDOPersistenceManager.deletePersistentAll(JDOPersistenceManager.java:814)
	at alpine.persistence.AbstractAlpineQueryManager.delete(AbstractAlpineQueryManager.java:484)
	at org.dependencytrack.persistence.PolicyQueryManager.reconcilePolicyViolations(PolicyQueryManager.java:168)
	at org.dependencytrack.persistence.QueryManager.reconcilePolicyViolations(QueryManager.java:533)
	at org.dependencytrack.policy.PolicyEngine.evaluate(PolicyEngine.java:98)
	at org.dependencytrack.policy.PolicyEngine.evaluate(PolicyEngine.java:69)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.performPolicyEvaluation(VulnerabilityAnalysisTask.java:130)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.inform(VulnerabilityAnalysisTask.java:76)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:101)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

	at org.datanucleus.store.rdbms.request.UpdateRequest.execute(UpdateRequest.java:508)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.updateObjectInTable(RDBMSPersistenceHandler.java:447)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.updateObject(RDBMSPersistenceHandler.java:421)
	at org.datanucleus.store.rdbms.mapping.java.PersistableMapping.preDelete(PersistableMapping.java:1104)
	at org.datanucleus.store.rdbms.request.DeleteRequest.execute(DeleteRequest.java:213)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObjectFromTable(RDBMSPersistenceHandler.java:534)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObject(RDBMSPersistenceHandler.java:506)
	at org.datanucleus.state.StateManagerImpl.internalDeletePersistent(StateManagerImpl.java:1260)
	at org.datanucleus.state.StateManagerImpl.deletePersistent(StateManagerImpl.java:5513)
	at org.datanucleus.ExecutionContextImpl.deleteObjectInternal(ExecutionContextImpl.java:2294)
	at org.datanucleus.ExecutionContextImpl.deleteObjectWork(ExecutionContextImpl.java:2220)
	at org.datanucleus.ExecutionContextImpl.deleteObjects(ExecutionContextImpl.java:2116)
	at org.datanucleus.api.jdo.JDOPersistenceManager.deletePersistentAll(JDOPersistenceManager.java:814)
	... 11 common frames omitted
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Cannot insert the value NULL into column 'POLICYVIOLATION_ID', table '15B-PRD-DependencyTrack.dbo.VIOLATIONANALYSIS'; column does not allow nulls. UPDATE fails.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:265)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1673)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:620)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:540)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7627)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3912)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:268)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:242)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeUpdate(SQLServerPreparedStatement.java:486)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at org.datanucleus.store.rdbms.SQLController.executeStatementUpdate(SQLController.java:430)
	at org.datanucleus.store.rdbms.request.UpdateRequest.execute(UpdateRequest.java:467)
	... 23 common frames omitted
```

### Steps to Reproduce

Unknown

### Expected Behavior

Not to error

### Dependency-Track Version

4.6.1

### Dependency-Track Distribution

Container Image

### Database Server

Microsoft SQL Server

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by BlythMeister** [comment]
This appears to happen when you had a policy violation which has analysis done and is then resolved (so needs to be deleted)

**Comment by mykter** [comment]
I think this is a duplicate of #1931 ?

**Comment by BlythMeister** [comment]
I would agree it is

**Comment by nscuro** [comment]
Closing as duplicate. However, this is fixed with #3228.
